kirill_tolstobrov/graph_generator.cpp

```diff
@@ -29,60 +36,136 @@ uni_cpp_practice::VertexId random_vertex_id(
-  const float probability_decreasement = 1.0 / depth;
-  float new_vertex_prob = 1.0;
-  for (int cur_depth = 0; cur_depth < depth; cur_depth++) {
-    if (graph.depths_map_.size() - 1 < cur_depth) {
-      return;
-    }
-    const auto& vertex_ids_at_depth = graph.depths_map_[cur_depth];
-    for (const VertexId cur_vertex_id : vertex_ids_at_depth) {
-      for (int k = 0; k < new_vertices_num; k++) {
-        if (random_bool(new_vertex_prob)) {
-          const auto new_vertex_id = graph.add_new_vertex();
-          graph.bind_vertices(cur_vertex_id, new_vertex_id);
+  using JobCallback = std::function<void()>;
+  auto jobs = std::list<JobCallback>();
+  std::mutex graph_mutex;
+
+  const VertexId first_vertex_id = 0;
+  std::atomic<int> done_jobs_number = 0;
+  for (int i = 0; i < new_vertices_num; i++) {
+    jobs.push_back(
+        [&graph, &graph_mutex, &done_jobs_number, &first_vertex_id, this]() {
+          generate_gray_branch(graph, 0, first_vertex_id, graph_mutex);
+          done_jobs_number++;
+        });
+  }
+
+  std::mutex jobs_mutex;
+  std::atomic<bool> should_terminate = false;
+  auto worker = [&jobs, &jobs_mutex, &should_terminate]() {
+    while (true) {
+      if (should_terminate) {
+        return;
+      }
+
+      const auto job_optional = [&jobs,
+                                 &jobs_mutex]() -> std::optional<JobCallback> {
+        const std::lock_guard lock(jobs_mutex);
+        if (!jobs.empty()) {
+          const auto job = jobs.back();
+          jobs.pop_back();
+          return job;
         }
+        return std::nullopt;
+      }();
+
+      if (job_optional.has_value()) {
+        const auto& job = job_optional.value();
+        job();
       }
     }
-    new_vertex_prob -= probability_decreasement;
+  };
+
+  const auto threads_count = std::min(MAX_THREADS_COUNT, new_vertices_num);
+  auto threads = std::vector<std::thread>();
+  threads.reserve(threads_count);
+
+  for (int i = 0; i < threads_count; i++) {
+    threads.push_back(std::thread(worker));
+  }
+
+  while (done_jobs_number < new_vertices_num) {
+  }
+
+  should_terminate = true;
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+void GraphGenerator::generate_gray_branch(Graph& graph,
+                                          const int cur_depth,
+                                          const VertexId& cur_vertex_id,
+                                          std::mutex& mutex) const {
+  if (cur_depth == params_.depth) {
+    return;
+  }
+
+  const float new_vertex_prob = 1.0 - (float)cur_depth / params_.depth;
+  int new_vertices_num = params_.new_vertices_num;
+  if (cur_vertex_id == 0) {
+    new_vertices_num = 1;
+  }
+  for (int i = 0; i < new_vertices_num; i++) {
+    if (random_bool(new_vertex_prob)) {
+      const std::lock_guard lock(mutex);
+      const auto new_vertex_id = graph.add_new_vertex();
+      graph.bind_vertices(cur_vertex_id, new_vertex_id);
+      lock.~lock_guard();
```

Comment:
lock_guard будет автоматически удален, когда закончится скоп, в котором он был объявлен. В этом его преимущество, что не нужно вручную управлять его памятью
```suggestion
      {
        const std::lock_guard lock(mutex);
        const auto new_vertex_id = graph.add_new_vertex();
        graph.bind_vertices(cur_vertex_id, new_vertex_id);
      }
```

Reply:
Там ведь потом вызов рекурсивный этой же функции с этим же мьютексом. Он ведь тогда, по идее, не разблокируется, и программа зависнет. У меня так и происходит: с этой строчкой работает, без нее прога виснет.

Reply:
Аа, все, понял